File: services/workspace_preview/static/modules.js
"use strict";
(() => {
 const $=id=>document.getElementById(id), api=(...args)=>window.wzosClock.api(...args);
 const safetyFields={competent_person:"Competent person / reviewer",tasks:"Job tasks",hazards:"Hazards",controls:"Controls",emergency_plan:"Emergency plan",ppe:"PPE"};
 for(const [key,label] of Object.entries(safetyFields)){const wrap=document.createElement('label');wrap.textContent=label;const input=document.createElement('textarea');input.id='safety-'+key;input.maxLength=key==='competent_person'?200:key==='ppe'?1000:key==='emergency_plan'?2000:4000;wrap.append(input);$('safety-fields').append(wrap);}
 let roster=[];
 const checks={tires:"Tires",fluids:"Fluids",brakes:"Brakes",ebrake:"Emergency brake",mirrors:"Mirrors",windows:"Windows"};
 for(const [key,label] of Object.entries(checks)){const wrap=document.createElement("label");wrap.textContent=label;const select=document.createElement("select");select.id="dvir-"+key;for(const [value,title] of Object.entries({not_checked:"Not checked",pass:"Pass (reported)",fail:"Fail (reported)"})){const option=document.createElement("option");option.value=value;option.textContent=title;select.append(option);}wrap.append(select);$("dvir-checks").append(wrap);}
 function template(){$("module-safety").hidden=kind!=="forms"||$("module-type").value!=="jsa";const enabled=kind==="forms"&&$("module-type").value==="dvir";$("module-inspection").hidden=!enabled;$("dvir-vehicle").required=enabled;}
 $("module-type").onchange=template;
 let kind="forms", current=null, dirty=false, saving=false, epoch=0, offset=0, total=0;
 const text=(tag,value)=>{const n=document.createElement(tag);n.textContent=value;return n;};
 const notice=value=>{$("module-notice").textContent=value;};
 window.wzosModulesCanLeave=()=>!saving&&(!dirty||confirm("Leave unsaved module changes?"));
 function localTime(value){if(!value)return "";const d=new Date(value);return new Date(d.getTime()-d.getTimezoneOffset()*60000).toISOString().slice(0,16);}
 function edit(row){
  let attachments=$("module-attachments");if(!attachments){attachments=text("div","");attachments.id="module-attachments";$("module-form").after(attachments);}window.wzosFiles.mount(attachments,"form",kind==="forms"?row?.id:null);
  $("module-type").value=row?.form_type||"incident";$("module-type").disabled=Boolean(row);
  const inspection=row?.inspection;$("dvir-vehicle").value=inspection?.vehicle_id||"";$("dvir-odometer").value=inspection?.odometer??"";$("dvir-trip").value=inspection?.trip_type||"pre-trip";$("dvir-defects").value=inspection?.defects||"";for(const key of Object.keys(checks))$("dvir-"+key).value=inspection?.[key]||"not_checked";template();
  for(const key of Object.keys(safetyFields))$("safety-"+key).value=row?.safety?.[key]||"";
  $("module-assignees").replaceChildren();for(const member of roster){const option=text("option",member.name+" · "+member.role);option.value=member.id;option.selected=(row?.assignees||[]).includes(member.id);$("module-assignees").append(option);}
  $("module-history").replaceChildren();if(row)loadHistory(row);
  current=row||{id:crypto.randomUUID(),version:0};dirty=false;$("module-form").hidden=false;
  for(const [id,key] of [["name","title"],["location","location"],["details","details"]])$("module-"+id).value=row?.[key]||"";
  const none=text("option","No linked work order");none.value="";$("module-order").replaceChildren(none);
  for(const order of window.wzosClock.getOrders()){const option=text("option",order.title);option.value=order.id;$("module-order").append(option);}
  $("module-order").value=row?.order_id||"";
  $("module-start").value=localTime(row?.start);$("module-end").value=localTime(row?.end);$("module-status").value=row?.status||"draft";
  $("module-version").textContent=row?`Saved revision ${row.version} · ${new Date(row.updated_at).toLocaleString()}`:"Unsaved draft";
  const readOnly=kind==="schedule"&&window.wzosClock.getSession()?.role!=="admin";
  for(const input of $("module-form").elements)input.disabled=readOnly;
  $("module-type").disabled=readOnly||Boolean(row);
 }
 async function refresh(){
  const generation=++epoch;
  try{const data=await api(`/api/modules/${kind}?offset=${offset}${kind==="schedule"&&$("module-day").value?"&day="+$("module-day").value:""}`);if(generation!==epoch)return;
   total=data.total;$("module-list").replaceChildren();$("module-new").hidden=!data.can_edit;
   for(const row of data.items){const button=text("button",`${row.title} · ${row.status}${row.start?" · "+new Date(row.start).toLocaleString():""}`);button.type="button";button.className="job";button.onclick=()=>{if(window.wzosModulesCanLeave())edit(row);};$("module-list").append(button);}
   if(!data.items.length)$("module-list").append(text("p","No saved records on this page."));
   $("module-page").textContent=`${total} records · ${data.items.length?offset+1:0}–${offset+data.items.length}`;
   $("module-prev").disabled=offset===0;$("module-next").disabled=offset+25>=total;
  }catch(error){if(generation===epoch)notice(error.message);}
 }
 async function show(view){
  if(!["forms","schedule"].includes(view))return;
  $("module-attachments")?.replaceChildren();const showGeneration=++epoch;kind=view;offset=0;dirty=false;current=null;$("module-form").hidden=true;notice("");
  $("module-title").textContent=kind==="forms"?"Forms hub":"Schedule management";
  $("module-description").textContent=kind==="forms"?"Incident, vehicle inspection and JSA planning drafts. Save, reopen and link to a work order. Official submissions and other templates are pending.":"Team schedule drafts. Admins edit; team members can read. Assignments use test members; overlapping active drafts are rejected. Saving does not dispatch or notify anyone.";
  $("module-day-label").hidden=kind!=="schedule";$("module-assignees-label").hidden=kind!=="schedule";$("module-history").replaceChildren();
  try{const result=await api("/api/modules/roster");if(showGeneration!==epoch)return;roster=result.items;}catch(error){if(showGeneration!==epoch)return;roster=[];notice(error.message);}
  $("module-type-label").hidden=kind!=="forms";
  $("module-times").hidden=kind!=="schedule";$("module-start").required=$("module-end").required=kind==="schedule";
  refresh();
 }
 for(const key of ["forms","schedule"])$(key+"-nav").onclick=()=>window.showWzosView(key);
 document.addEventListener("wzos:view",event=>show(event.detail));
 document.addEventListener("wzos:session",()=>{$("module-attachments")?.replaceChildren();epoch++;dirty=false;current=null;$("module-list").replaceChildren();$("module-form").hidden=true;if(!$("modules-view").hidden)show(kind);});
 $("module-day").onchange=()=>{offset=0;refresh();};
 async function loadHistory(row){const selectedKind=kind;try{const data=await api(`/api/modules/${kind}/${row.id}/history`);if(current?.id!==row.id||selectedKind!==kind)return;$("module-history").append(text("h3","Saved revision history"));for(const entry of data.items){const d=text("details","");d.append(text("summary",`Revision ${entry.version} · ${new Date(entry.saved_at).toLocaleString()}`),text("p",entry.record.title),text("p",entry.record.details||"No notes"));for(const [key,value] of Object.entries(entry.record.safety||{}))d.append(text("p",`${safetyFields[key]}: ${value||"Not recorded"}`));$("module-history").append(d);}}catch(error){notice(error.message);}}
 $("module-new").onclick=()=>{if(window.wzosModulesCanLeave())edit(null);};
 $("module-refresh").onclick=()=>{if(window.wzosModulesCanLeave()){dirty=false;$("module-form").hidden=true;refresh();}};
 $("module-prev").onclick=()=>{if(window.wzosModulesCanLeave()){offset=Math.max(0,offset-25);refresh();}};
 $("module-next").onclick=()=>{if(window.wzosModulesCanLeave()){offset+=25;refresh();}};
 $("module-form").oninput=()=>{dirty=true;};
 window.addEventListener("beforeunload",event=>{if(dirty||saving){event.preventDefault();event.returnValue="";}});
 $("module-form").onsubmit=async event=>{
  event.preventDefault();if(saving)return;
  const generation=epoch, record=current;
  const body={request_id:record.id,expected_version:record.version,title:$("module-name").value,location:$("module-location").value,details:$("module-details").value,order_id:$("module-order").value||null,status:$("module-status").value,start:null,end:null};
  if(kind==="forms"){body.form_type=$("module-type").value;if(body.form_type==="dvir"){body.inspection={vehicle_id:$("dvir-vehicle").value,odometer:$("dvir-odometer").value===""?null:Number($("dvir-odometer").value),trip_type:$("dvir-trip").value,defects:$("dvir-defects").value};for(const key of Object.keys(checks))body.inspection[key]=$("dvir-"+key).value;if(Object.keys(checks).some(key=>body.inspection[key]==="fail")&&!body.inspection.defects.trim()){notice("Describe the failed inspection items in Defects.");return;}}}
  if(kind==="forms"&&body.form_type==="jsa"){body.safety={};for(const key of Object.keys(safetyFields))body.safety[key]=$("safety-"+key).value;}
  if(kind==="schedule"){body.assignees=Array.from($("module-assignees").selectedOptions,o=>o.value);body.start=new Date($("module-start").value).toISOString();body.end=new Date($("module-end").value).toISOString();if(body.end<=body.start){notice("End must be after start.");return;}}
  saving=true;for(const input of $("module-form").elements)input.disabled=true;$("identity").disabled=true;
  try{const saved=await api(`/api/modules/${kind}/${record.id}`,{method:"PUT",body:JSON.stringify(body)});if(generation!==epoch)return;edit(saved);notice("Draft saved. No submission or notification was sent.");await refresh();}
  catch(error){notice(error.message);}finally{saving=false;for(const input of $("module-form").elements)input.disabled=false;$("identity").disabled=false;$("module-type").disabled=Boolean(current?.version);}
 };
})();

File: services/workspace_preview/modules.py
"""Scoped synthetic scheduling and incident drafts adapted from recovered Core fields."""
import json
from datetime import date, datetime, timezone
from typing import Literal
from uuid import UUID
from fastapi import HTTPException, Request, Query
from pydantic import BaseModel, ConfigDict, Field, model_validator

class VehicleInspection(BaseModel):
    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)
    vehicle_id: str = Field(min_length=1, max_length=120)
    odometer: int | None = Field(default=None, ge=0, le=10000000, strict=True)
    trip_type: Literal["pre-trip", "post-trip"] = "pre-trip"
    tires: Literal["not_checked", "pass", "fail"] = "not_checked"
    fluids: Literal["not_checked", "pass", "fail"] = "not_checked"
    brakes: Literal["not_checked", "pass", "fail"] = "not_checked"
    ebrake: Literal["not_checked", "pass", "fail"] = "not_checked"
    mirrors: Literal["not_checked", "pass", "fail"] = "not_checked"
    windows: Literal["not_checked", "pass", "fail"] = "not_checked"
    defects: str = Field(default="", max_length=4000)

    @model_validator(mode="after")
    def defect_details(self):
        if any(getattr(self, key) == "fail" for key in ("tires", "fluids", "brakes", "ebrake", "mirrors", "windows")) and not self.defects:
            raise ValueError("Describe any failed inspection item in Defects")
        return self


class SafetyWorksheet(BaseModel):
    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)
    competent_person: str = Field(default="", max_length=200)
    tasks: str = Field(default="", max_length=4000)
    hazards: str = Field(default="", max_length=4000)
    controls: str = Field(default="", max_length=4000)
    emergency_plan: str = Field(default="", max_length=2000)
    ppe: str = Field(default="", max_length=1000)

class ModuleRecord(BaseModel):
    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)
    request_id: UUID
    expected_version: int = Field(default=0, ge=0, strict=True)
    form_type: Literal["incident", "dvir", "jsa"] = "incident"
    inspection: VehicleInspection | None = None
    safety: SafetyWorksheet | None = None
    assignees: list[str] = Field(default_factory=list, max_length=50)
    title: str = Field(min_length=1, max_length=120)
    location: str = Field(default="", max_length=500)
    details: str = Field(default="", max_length=4000)
    order_id: str | None = Field(default=None, max_length=100)
    start: datetime | None = None
    end: datetime | None = None
    status: Literal["draft", "cancelled"] = "draft"

    @model_validator(mode="after")
    def times(self):
        for value in (self.start, self.end):
            if value and (value.tzinfo is None or value.utcoffset() is None):
                raise ValueError("Schedule times require a timezone")
        if self.start and self.end and self.end <= self.start:
            raise ValueError("End must be after start")
        return self


def register(app, connect, actor, permitted_row, actors, synthetic=True):
    with connect() as db:
        db.execute("""CREATE TABLE IF NOT EXISTS module_records (
            id TEXT NOT NULL, kind TEXT NOT NULL, organization_id TEXT NOT NULL,
            owner_id TEXT NOT NULL, version INTEGER NOT NULL, payload TEXT NOT NULL,
            updated_at TEXT NOT NULL, PRIMARY KEY(organization_id, kind, id))""")

        db.execute("""CREATE TABLE IF NOT EXISTS module_revisions (
            organization_id TEXT NOT NULL, kind TEXT NOT NULL, id TEXT NOT NULL,
            version INTEGER NOT NULL, payload TEXT NOT NULL, author_id TEXT NOT NULL,
            saved_at TEXT NOT NULL, PRIMARY KEY(organization_id,kind,id,version))""")

    @app.get("/api/modules/roster")
    def roster(request: Request):
        selected=actor(request)
        return {"items":[{"id":a["id"],"name":a["name"],"role":a["role"]} for a in actors(selected).values() if a["organization_id"]==selected["organization_id"]], "synthetic":synthetic}

    def scope(kind, selected):
        clause = "organization_id=? AND kind=?"
        params = [selected["organization_id"], kind]
        if kind == "forms" and selected["role"] != "admin":
            clause += " AND owner_id=?"
            params.append(selected["id"])
        return clause, params

    def unpack(row):
        return {**json.loads(row["payload"]), "id": row["id"], "version": row["version"],
                "updated_at": row["updated_at"], "owner_id": row["owner_id"]}

    @app.get("/api/modules/{kind}")
    def listing(kind: Literal["forms", "schedule"], request: Request,
                offset: int = Query(0, ge=0), limit: int = Query(25, ge=1, le=50), day: date | None = None):
        selected = actor(request)
        clause, params = scope(kind, selected)
        if day and kind == "schedule":
            clause += " AND substr(json_extract(payload, '$.start'),1,10)=?"
            params.append(day.isoformat())
        with connect() as db:
            total = db.execute("SELECT COUNT(*) FROM module_records WHERE " + clause, params).fetchone()[0]
            rows = db.execute("SELECT * FROM module_records WHERE " + clause +
                              " ORDER BY updated_at DESC, id LIMIT ? OFFSET ?", [*params, limit, offset]).fetchall()
        return {"items": [unpack(row) for row in rows], "total": total, "offset": offset,
                "can_edit": kind == "forms" or selected["role"] == "admin"}

    @app.put("/api/modules/{kind}/{record_id}")
    def save(kind: Literal["forms", "schedule"], record_id: UUID, payload: ModuleRecord, request: Request):
        selected = actor(request)
        if kind == "schedule":
            if selected["role"] != "admin":
                raise HTTPException(403, "Only admins can edit the team schedule")
            if not payload.start or not payload.end:
                raise HTTPException(422, "Start and end times are required")
        elif payload.form_type == "dvir" and payload.inspection is None:
            raise HTTPException(422, "Vehicle inspection details are required")
        if kind == "forms" and payload.form_type == "jsa" and payload.safety is None:
            raise HTTPException(422, "Safety worksheet fields are required")
        if payload.form_type != "jsa" and payload.safety is not None:
            raise HTTPException(422, "Safety fields belong to the JSA worksheet")
        if payload.form_type != "dvir" and payload.inspection is not None:
            raise HTTPException(422, "Incident drafts cannot contain vehicle inspection fields")
        if kind == "schedule" and (payload.inspection is not None or payload.form_type != "incident"):
            raise HTTPException(422, "Vehicle inspections belong in Forms hub")
        if kind == "forms" and (payload.start or payload.end):
            raise HTTPException(422, "Incident drafts do not use schedule times")
        allowed={a["id"] for a in actors(selected).values() if a["organization_id"]==selected["organization_id"]}
        if len(set(payload.assignees)) != len(payload.assignees) or not set(payload.assignees)<=allowed or (kind=="forms" and payload.assignees):
            raise HTTPException(422, "Assignments must be distinct members of this organization on a schedule")
        packed = payload.model_dump(mode="json", exclude={"expected_version", "request_id"})
        for key in ("start", "end"):
            if getattr(payload, key):
                packed[key] = getattr(payload, key).astimezone(timezone.utc).isoformat()
        serialized = json.dumps(packed, sort_keys=True)
        with connect() as db:
            db.execute("BEGIN IMMEDIATE")
            clause, params = scope(kind, selected)
            row = db.execute("SELECT * FROM module_records WHERE " + clause + " AND id=?", [*params, str(record_id)]).fetchone()
            existing = db.execute("SELECT 1 FROM module_records WHERE organization_id=? AND kind=? AND id=?",
                                  (selected["organization_id"], kind, str(record_id))).fetchone()
            if existing and not row:
                raise HTTPException(404, "Record not found")
            if row and kind == "forms" and json.loads(row["payload"]).get("form_type", "incident") != payload.form_type:
                raise HTTPException(409, "Create a new draft to use a different form template")
            if payload.order_id:
                permitted_row(db, payload.order_id, selected)
            version = row["version"] if row else 0
            # Retrying the same saved content is safe; stale conflicting edits are not.
            if row and row["payload"] == serialized and payload.expected_version == version - 1:
                return unpack(row)
            if version != payload.expected_version:
                raise HTTPException(409, "Record changed. Reload before editing")
            if kind=="schedule" and payload.status!="cancelled" and payload.assignees:
                others=db.execute("SELECT id,payload FROM module_records WHERE organization_id=? AND kind='schedule' AND id<>?",(selected["organization_id"],str(record_id))).fetchall()
                for other in others:
                    event=json.loads(other["payload"])
                    if event.get("status")!="cancelled" and set(event.get("assignees",[])) & set(payload.assignees) and event.get("start") and event.get("end"):
                        if datetime.fromisoformat(event['start']) < payload.end and datetime.fromisoformat(event['end']) > payload.start:
                            raise HTTPException(409,"An assigned member has an overlapping schedule draft")
            if row:
                db.execute("INSERT INTO module_revisions VALUES (?,?,?,?,?,?,?) ON CONFLICT DO NOTHING",(selected["organization_id"],kind,str(record_id),row['version'],row['payload'],row['owner_id'],row['updated_at']))
            db.execute("INSERT INTO module_revisions VALUES (?,?,?,?,?,?,?)",(selected["organization_id"],kind,str(record_id),version+1,serialized,selected['id'],datetime.now(timezone.utc).isoformat()))
            db.execute("INSERT INTO module_records VALUES (?,?,?,?,?,?,?) ON CONFLICT(organization_id,kind,id) DO UPDATE SET version=excluded.version,payload=excluded.payload,updated_at=excluded.updated_at",
                       (str(record_id), kind, selected["organization_id"], row["owner_id"] if row else selected["id"],
                        version + 1, serialized, datetime.now(timezone.utc).isoformat()))
            return unpack(db.execute("SELECT * FROM module_records WHERE organization_id=? AND kind=? AND id=?",
                                     (selected["organization_id"], kind, str(record_id))).fetchone())

    @app.get('/api/modules/{kind}/{record_id}/history')
    def history(kind: Literal['forms','schedule'], record_id: UUID, request: Request, offset: int=Query(0,ge=0)):
        selected=actor(request);clause,params=scope(kind,selected)
        with connect() as db:
            current=db.execute('SELECT * FROM module_records WHERE '+clause+' AND id=?',(*params,str(record_id))).fetchone()
            if not current: raise HTTPException(404,'Record not found')
            rows=db.execute('SELECT * FROM module_revisions WHERE organization_id=? AND kind=? AND id=? ORDER BY version DESC LIMIT 20 OFFSET ?', (selected['organization_id'],kind,str(record_id),offset)).fetchall()
        return {"items":[{"version":r['version'],"saved_at":r['saved_at'],"author_id":r['author_id'],"record":json.loads(r['payload'])} for r in rows]}
